In [81]:
import numpy as np
import pandas as pd

In [82]:
df = pd.read_csv('/content/drive/MyDrive/Curso ML/heart_tratado.csv', sep=';', encoding='utf-8')

In [83]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140        289.0          0     Normal    172   
1   49   F           NAP        160        180.0          0     Normal    156   
2   37   M           ATA        130        283.0          0         ST     98   
3   48   F           ASY        138        214.0          0     Normal    108   
4   54   M           NAP        150        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [84]:
df.shape

(917, 12)

In [85]:
df.dtypes

Age                 int64
Sex                object
ChestPainType      object
RestingBP           int64
Cholesterol       float64
FastingBS           int64
RestingECG         object
MaxHR               int64
ExerciseAngina     object
Oldpeak           float64
ST_Slope           object
HeartDisease        int64
dtype: object

In [86]:
#Análise das escalas do Atributos
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  917.000000  917.000000   917.000000  917.000000  917.000000   
mean    53.509269  132.540894   244.635389    0.233370  136.789531   
std      9.437636   17.999749    53.347125    0.423206   25.467129   
min     28.000000   80.000000    85.000000    0.000000   60.000000   
25%     47.000000  120.000000   214.000000    0.000000  120.000000   
50%     54.000000  130.000000   244.635389    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  917.000000    917.000000  
mean     0.886696      0.552890  
std      1.066960      0.497466  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

#Legenda
Age = idade (anos)

Sex = sexo (0=M; 1=F)

Chest Pain Type = tipo de dor no peito (0=TA:angina típica; 1=ATA:angina atípica; 2=NAP:dor não anginosa; 3=ASY:assintomático)

Resting BP = pressão sanguínea em repouso (mmHg)

Cholesterol = colesterol sérico (mg/dl)

Fasting BS = açúcar no sangue em jejum (mg/dl)
          0: Fasting BS < 120 mg/dl (não diabético)
          1: Fasting BS >= 120 mg/dl, (diabético)

Resting ECG = eletrocardiograma em repouso (0=Normal; 1=ST:Anormalidade da onda ST-T; 2=LVH: Hipertrofia ventricular esquerda)

Max HR = frequência cardíaca máxima

Exercise Angina = Angina induzida por exercício (0=Não; 1=Sim)

Old Peak = Depressão de ST induzida por exercício em relação ao repouso

ST_Slope = Inclinação do segmento ST (0=UP; 1=Flat; 2=Down)

Heart Disease = Doença cardíaca (0=Não possui doença cardíaca; 1=Possui doença cardíaca)


In [87]:
# Atributos previsores e Alvo
previsores = df.iloc[:,0:11].values

In [88]:
previsores

array([[40, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 'F', 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 'M', 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 'F', 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 'M', 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [89]:
previsores.shape

(917, 11)

In [90]:
alvo = df.iloc[:, 11].values

In [ ]:
alvo

In [92]:
alvo.shape

(917,)

In [93]:
#LabelEncoder
from sklearn.preprocessing import LabelEncoder

In [94]:
previsoresLabelEncoder = df.iloc[:, 0:11].values
previsoresLabelEncoder

array([[40, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 'F', 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 'M', 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 'F', 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 'M', 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [95]:
previsoresLabelEncoder[:,1] = LabelEncoder().fit_transform(previsores[:,1])
previsoresLabelEncoder[:,2] = LabelEncoder().fit_transform(previsores[:,2])
previsoresLabelEncoder[:,6] = LabelEncoder().fit_transform(previsores[:,6])
previsoresLabelEncoder[:,8] = LabelEncoder().fit_transform(previsores[:,8])
previsoresLabelEncoder[:,10] = LabelEncoder().fit_transform(previsores[:,10])

In [96]:
previsoresLabelEncoder

array([[40, 1, 1, ..., 0, 0.0, 2],
       [49, 0, 2, ..., 0, 1.0, 1],
       [37, 1, 1, ..., 0, 0.0, 2],
       ...,
       [57, 1, 0, ..., 1, 1.2, 1],
       [57, 0, 1, ..., 0, 0.0, 1],
       [38, 1, 2, ..., 0, 0.0, 2]], dtype=object)

In [97]:
previsoresLabelEncoder.shape

(917, 11)

In [98]:
# OneHotEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

In [99]:
previsoresOneHotEncoder = ColumnTransformer(transformers=[('OneHot', OneHotEncoder(), [1,2,6,8,10])],
                                remainder = 'passthrough').fit_transform(previsoresLabelEncoder)

In [100]:
previsoresOneHotEncoder

array([[0.0, 1.0, 0.0, ..., 0, 172, 0.0],
       [1.0, 0.0, 0.0, ..., 0, 156, 1.0],
       [0.0, 1.0, 0.0, ..., 0, 98, 0.0],
       ...,
       [0.0, 1.0, 1.0, ..., 0, 115, 1.2],
       [1.0, 0.0, 0.0, ..., 0, 174, 0.0],
       [0.0, 1.0, 0.0, ..., 0, 173, 0.0]], dtype=object)

In [101]:
previsoresOneHotEncoder.shape

(917, 20)

In [102]:
previsoresOneHotEncoder = pd.DataFrame(previsoresOneHotEncoder)
previsoresOneHotEncoder.head()

0    1    2    3    4    5    6    7    8    9    10   11   12   13  14  \
0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  40   
1  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0  49   
2  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0  37   
3  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  48   
4  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  54   

    15     16 17   18   19  
0  140  289.0  0  172  0.0  
1  160  180.0  0  156  1.0  
2  130  283.0  0   98  0.0  
3  138  214.0  0  108  1.5  
4  150  195.0  0  122  0.0

In [103]:
#Realizar a padronização
from sklearn.preprocessing import StandardScaler

previsores_padronizado = StandardScaler().fit_transform(previsoresOneHotEncoder)

In [104]:
previsores_padronizado

array([[-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
         1.38333943, -0.83150225],
       [ 1.9368261 , -1.9368261 , -1.08542493, ..., -0.55173333,
         0.75473573,  0.10625149],
       [-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
        -1.52395266, -0.83150225],
       ...,
       [-0.51630861,  0.51630861,  0.92129817, ..., -0.55173333,
        -0.85606123,  0.29380223],
       [ 1.9368261 , -1.9368261 , -1.08542493, ..., -0.55173333,
         1.46191489, -0.83150225],
       [-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
         1.42262716, -0.83150225]])

In [105]:
previsores_padronizado= pd.DataFrame(previsores_padronizado)
previsores_padronizado

0         1         2         3         4         5         6   \
0   -0.516309  0.516309 -1.085425  2.073784 -0.531524 -0.229810 -0.507826   
1    1.936826 -1.936826 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   
2   -0.516309  0.516309 -1.085425  2.073784 -0.531524 -0.229810 -0.507826   
3    1.936826 -1.936826  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
4   -0.516309  0.516309 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   
..        ...       ...       ...       ...       ...       ...       ...   
912 -0.516309  0.516309 -1.085425 -0.482210 -0.531524  4.351412 -0.507826   
913 -0.516309  0.516309  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
914 -0.516309  0.516309  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
915  1.936826 -1.936826 -1.085425  2.073784 -0.531524 -0.229810  1.969177   
916 -0.516309  0.516309 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   

           7         8         9         10        11        12        13  \
0    0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
1    0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
2   -1.226974  2.037569  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
3    0.815013 -0.490781 -1.213136  1.213136 -0.271607  0.998910 -0.869888   
4    0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
..        ...       ...       ...       ...       ...       ...       ...   
912  0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
913  0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
914  0.815013 -0.490781 -1.213136  1.213136 -0.271607  0.998910 -0.869888   
915 -1.226974 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
916  0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   

           14        15        16        17        18        19  
0   -1.432206  0.414627  0.832075 -0.551733  1.383339 -0.831502  
1   -0.478057  1.526360 -1.212261 -0.551733  0.754736  0.106251  
2   -1.750256 -0.141240  0.719543 -0.551733 -1.523953 -0.831502  
3   -0.584074  0.303453 -0.574578 -0.551733 -1.131075  0.575128  
4    0.052026  0.970493 -0.930931 -0.551733 -0.581047 -0.831502  
..        ...       ...       ...       ...       ...       ...  
912 -0.902124 -1.252973  0.363191 -0.551733 -0.188170  0.293802  
913  1.536257  0.636973 -0.968441  1.812470  0.165420  2.356860  
914  0.370075 -0.141240 -2.131275 -0.551733 -0.856061  0.293802  
915  0.370075 -0.141240 -0.161960 -0.551733  1.461915 -0.831502  
916 -1.644239  0.303453 -1.306038 -0.551733  1.422627 -0.831502  

[917 rows x 20 columns]

In [106]:
previsores_padronizado.describe()

0             1             2             3             4   \
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02   
mean   1.084798e-16 -1.472226e-16  1.937139e-17 -3.874279e-17  3.874279e-17   
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00   
min   -5.163086e-01 -1.936826e+00 -1.085425e+00 -4.822104e-01 -5.315237e-01   
25%   -5.163086e-01  5.163086e-01 -1.085425e+00 -4.822104e-01 -5.315237e-01   
50%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
75%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
max    1.936826e+00  5.163086e-01  9.212982e-01  2.073784e+00  1.881384e+00   

                 5           6             7             8             9   \
count  9.170000e+02  917.000000  9.170000e+02  9.170000e+02  9.170000e+02   
mean   6.973702e-17    0.000000 -9.298269e-17  1.549712e-17 -4.261707e-17   
std    1.000546e+00    1.000546  1.000546e+00  1.000546e+00  1.000546e+00   
min   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
25%   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
50%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
75%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
max    4.351412e+00    1.969177  8.150134e-01  2.037569e+00  8.243101e-01   

                 10            11          12            13            14  \
count  9.170000e+02  9.170000e+02  917.000000  9.170000e+02  9.170000e+02   
mean   4.261707e-17  8.523413e-17    0.000000 -3.874279e-17  1.859654e-16   
std    1.000546e+00  1.000546e+00    1.000546  1.000546e+00  1.000546e+00   
min   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -2.704405e+00   
25%   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -6.900904e-01   
50%   -8.243101e-01 -2.716072e-01    0.998910 -8.698879e-01  5.202558e-02   
75%    1.213136e+00 -2.716072e-01    0.998910  1.149573e+00  6.881250e-01   
max    1.213136e+00  3.681787e+00    0.998910  1.149573e+00  2.490407e+00   

                 15            16            17            18            19  
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  
mean   7.884157e-16  3.014189e-15 -1.549712e-17 -5.114048e-16 -1.859654e-16  
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  
min   -2.920572e+00 -2.994023e+00 -5.517333e-01 -3.016886e+00 -3.269662e+00  
25%   -6.971063e-01 -5.745784e-01 -5.517333e-01 -6.596226e-01 -8.315022e-01  
50%   -1.412398e-01  3.198360e-15 -5.517333e-01  4.755658e-02 -2.688500e-01  
75%    4.146267e-01  4.194568e-01 -5.517333e-01  7.547357e-01  5.751284e-01  
max    3.749826e+00  6.721265e+00  1.812470e+00  2.561971e+00  4.982571e+00

In [107]:
previsoresOneHotEncoder.describe()

0      1      2      3      4      5      6      7      8      9   \
count   917.0  917.0  917.0  917.0  917.0  917.0  917.0  917.0  917.0  917.0   
unique    2.0    2.0    2.0    2.0    2.0    2.0    2.0    2.0    2.0    2.0   
top       0.0    1.0    1.0    0.0    0.0    0.0    0.0    1.0    0.0    1.0   
freq    724.0  724.0  496.0  744.0  715.0  871.0  729.0  551.0  739.0  546.0   

           10     11     12     13   14   15          16   17   18     19  
count   917.0  917.0  917.0  917.0  917  917  917.000000  917  917  917.0  
unique    2.0    2.0    2.0    2.0   50   66  222.000000    2  119   53.0  
top       0.0    0.0    1.0    0.0   54  120  244.635389    0  150    0.0  
freq    546.0  854.0  459.0  522.0   51  132  171.000000  703   43  368.0

In [241]:
previsoresPre_processados = previsoresOneHotEncoder.apply(lambda col: col.astype(float) if col.dtype == 'object' else col)

In [242]:
previsoresPre_processados.dtypes

0     float64
1     float64
2     float64
3     float64
4     float64
5     float64
6     float64
7     float64
8     float64
9     float64
10    float64
11    float64
12    float64
13    float64
14    float64
15    float64
16    float64
17    float64
18    float64
19    float64
dtype: object

In [243]:
#Separação Treino e teste

from sklearn.model_selection import train_test_split

In [244]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsoresPre_processados, alvo, test_size = 0.3, random_state = 0)

In [245]:
x_treino.shape

(641, 20)

In [246]:
x_teste.shape

(276, 20)

In [247]:
y_treino.shape

(641,)

In [248]:
y_teste.shape

(276,)

In [249]:
#Naive Beyes
from sklearn.naive_bayes import GaussianNB

In [250]:
naive = GaussianNB()
naive.fit(x_treino, y_treino)

GaussianNB()

In [251]:
# Avaliação do Algoritmo
previsoes_naive = naive.predict(x_teste)
previsoes_naive

array([1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1])

In [252]:
y_teste

array([1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1,
       1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1])

In [253]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [254]:
accuracy_score(y_teste, previsoes_naive)

0.8478260869565217

In [255]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_naive)*100.0))

Acurácia: 84.78%


In [256]:
confusion_matrix(y_teste, previsoes_naive)

array([[100,  21],
       [ 21, 134]])

In [257]:
print(classification_report(y_teste, previsoes_naive))

              precision    recall  f1-score   support

           0       0.83      0.83      0.83       121
           1       0.86      0.86      0.86       155

    accuracy                           0.85       276
   macro avg       0.85      0.85      0.85       276
weighted avg       0.85      0.85      0.85       276



In [ ]:
#Análise de dados Treino

previsoes_treino = naive.predict(x_treino)
previsoes_treino

In [259]:
accuracy_score(y_treino, previsoes_treino)

0.8611544461778471

In [260]:
confusion_matrix(y_treino, previsoes_treino)

array([[248,  41],
       [ 48, 304]])

In [261]:
# Validação Cruzada

from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [262]:
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [263]:
modelo = GaussianNB()
resultado = cross_val_score(modelo, previsoresPre_processados, alvo, cv=kfold)

In [264]:
print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 85.17%


In [265]:
#Máquinas de Vetores de Suporte (SVM)
from sklearn.svm import SVC

In [266]:
svm = SVC(kernel='rbf', random_state=1, C = 600)
svm.fit(x_treino, y_treino)

SVC(C=600, random_state=1)

In [ ]:
previsoes_svm = svm.predict(x_teste)
previsoes_svm

In [ ]:
y_teste

In [269]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_svm)*100.0))

Acurácia: 85.51%


In [270]:
confusion_matrix(y_teste, previsoes_svm)

array([[ 99,  22],
       [ 18, 137]])

In [271]:
print(classification_report(y_teste, previsoes_svm))

              precision    recall  f1-score   support

           0       0.85      0.82      0.83       121
           1       0.86      0.88      0.87       155

    accuracy                           0.86       276
   macro avg       0.85      0.85      0.85       276
weighted avg       0.85      0.86      0.85       276



In [ ]:
#Análise Treino

previsoes_treino = svm.predict(x_treino)
previsoes_treino

In [273]:
accuracy_score(y_treino, previsoes_treino)

0.8689547581903276

In [274]:
confusion_matrix(y_treino, previsoes_treino)

array([[234,  55],
       [ 29, 323]])

In [275]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = SVC(kernel='rbf', random_state=1, C = 2)
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 85.72%


In [276]:
# Regressão Logística. É um algoritmo de CLASSIFICAÇÃO que utiliza conceitos de regressão linear.
from sklearn.linear_model import LogisticRegression

In [ ]:
logistica = LogisticRegression(random_state=1, max_iter=100, penalty="l2",
                               tol=0.0001, C=1, solver='lbfgs')
logistica.fit(x_treino, y_treino)

In [278]:
logistica.intercept_

array([-0.04306197])

In [ ]:
logistica.coef_

In [ ]:
previsoes_logistica = logistica.predict(x_teste)
previsoes_logistica

In [281]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_logistica)*100.0))

print(confusion_matrix(y_teste, previsoes_logistica))

print(classification_report(y_teste, previsoes_logistica))

Acurácia: 86.23%
[[102  19]
 [ 19 136]]
              precision    recall  f1-score   support

           0       0.84      0.84      0.84       121
           1       0.88      0.88      0.88       155

    accuracy                           0.86       276
   macro avg       0.86      0.86      0.86       276
weighted avg       0.86      0.86      0.86       276



In [282]:
#Análise dados Treino (Regressão Logística)
previsoes_treino = logistica.predict(x_treino)

print(accuracy_score(y_treino,previsoes_treino))

confusion_matrix(y_treino, previsoes_treino)

0.8642745709828393


array([[238,  51],
       [ 36, 316]])

In [283]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = LogisticRegression(random_state=1, max_iter=100, penalty="l2",
                               tol=0.0001, C=1, solver='newton-cg')
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 85.83%


In [284]:
# Aprendizagem baseada em instâncias (KNN)
from sklearn.neighbors import KNeighborsClassifier

In [285]:
knn = KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=1)
knn.fit(x_treino, y_treino)

KNeighborsClassifier(p=1)

In [286]:
previsoes_knn = knn.predict(x_teste)

In [ ]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_knn)*100.0))

print(confusion_matrix(y_teste, previsoes_knn))

print(classification_report(y_teste, previsoes_knn))

In [288]:
#Análise dados Treino (KNN)
previsoes_treino = knn.predict(x_treino)

print(accuracy_score(y_treino,previsoes_treino))

confusion_matrix(y_treino, previsoes_treino)

0.828393135725429


array([[221,  68],
       [ 42, 310]])

In [290]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2)
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 85.83%


In [291]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier

In [ ]:
arvore = DecisionTreeClassifier(criterion='gini', random_state = 0, max_depth=3)
arvore.fit(x_treino, y_treino)
previsoes_arvore = arvore.predict(x_teste)
previsoes_arvore

In [293]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_arvore)*100.0))

print(confusion_matrix(y_teste, previsoes_arvore))

print(classification_report(y_teste, previsoes_arvore))

Acurácia: 83.70%
[[ 98  23]
 [ 22 133]]
              precision    recall  f1-score   support

           0       0.82      0.81      0.81       121
           1       0.85      0.86      0.86       155

    accuracy                           0.84       276
   macro avg       0.83      0.83      0.83       276
weighted avg       0.84      0.84      0.84       276



In [294]:
#Análise dados Treino (Decision Tree)
previsoes_treino = arvore.predict(x_treino)

print(accuracy_score(y_treino,previsoes_treino))

confusion_matrix(y_treino, previsoes_treino)

0.8658346333853354


array([[233,  56],
       [ 30, 322]])

In [295]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = DecisionTreeClassifier(criterion='gini', random_state = 0, max_depth=3)
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 83.20%


In [296]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

In [ ]:
random = RandomForestClassifier(n_estimators=150, criterion='entropy', random_state=0, max_depth=6)
random.fit(x_treino,y_treino)
previsoes_random = random.predict(x_teste)
previsoes_random

In [298]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_random)*100.0))

print(confusion_matrix(y_teste, previsoes_random))

print(classification_report(y_teste, previsoes_random))

Acurácia: 83.70%
[[ 99  22]
 [ 23 132]]
              precision    recall  f1-score   support

           0       0.81      0.82      0.81       121
           1       0.86      0.85      0.85       155

    accuracy                           0.84       276
   macro avg       0.83      0.83      0.83       276
weighted avg       0.84      0.84      0.84       276



In [299]:
#Análise dados Treino (random Forest)
previsoes_treino = random.predict(x_treino)

print(accuracy_score(y_treino,previsoes_treino))

confusion_matrix(y_treino, previsoes_treino)

0.9266770670826833


array([[260,  29],
       [ 18, 334]])

In [300]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = RandomForestClassifier(n_estimators=150, criterion='entropy', random_state=0, max_depth=6)
resultado = cross_val_score(modelo, previsoresPre_processados, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 86.48%


In [301]:
# XGBoost
from xgboost import XGBClassifier

In [302]:
xg = XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)
xg.fit(x_treino, y_treino)
previsoes_xg = xg.predict(x_teste)
previsoes_xg

array([1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1])

In [303]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_xg)*100.0))

print(confusion_matrix(y_teste, previsoes_xg))

print(classification_report(y_teste, previsoes_xg))

Acurácia: 86.59%
[[102  19]
 [ 18 137]]
              precision    recall  f1-score   support

           0       0.85      0.84      0.85       121
           1       0.88      0.88      0.88       155

    accuracy                           0.87       276
   macro avg       0.86      0.86      0.86       276
weighted avg       0.87      0.87      0.87       276



In [304]:
#Análise dados Treino (XGBoost)
previsoes_treino = xg.predict(x_treino)

print(accuracy_score(y_treino,previsoes_treino))

confusion_matrix(y_treino, previsoes_treino)

0.9157566302652106


array([[254,  35],
       [ 19, 333]])

In [305]:
#Validação Cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

modelo = XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

print("Acurácia Média: %.2f%%" % (resultado.mean()* 100.0))

Acurácia Média: 86.59%


In [307]:
#LightGBM
import lightgbm as lgb

In [308]:
#Dataset para treino
dataset = lgb.Dataset(x_treino, label=y_treino)

In [309]:
#Parâmetros
parametros = {'num_leaves': 250,
              'objective':'binary',
              'max_depth':2,
              'learning_rate':.05,
              'max_bin':100}

In [310]:
lgbm=lgb.train(parametros,dataset,num_boost_round=200)

[LightGBM] [Info] Number of positive: 352, number of negative: 289
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000260 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 343
[LightGBM] [Info] Number of data points in the train set: 641, number of used features: 20
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.549142 -> initscore=0.197204
[LightGBM] [Info] Start training from score 0.197204
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -in

In [311]:
#Marcação do tempo de execução
from datetime import datetime
inicio=datetime.now()
lgbm=lgb.train(parametros,dataset)
fim=datetime.now()

tempo= fim - inicio
tempo

[LightGBM] [Info] Number of positive: 352, number of negative: 289
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000130 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 343
[LightGBM] [Info] Number of data points in the train set: 641, number of used features: 20
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.549142 -> initscore=0.197204
[LightGBM] [Info] Start training from score 0.197204


datetime.timedelta(microseconds=45647)

In [ ]:
previsoes_lgbm = lgbm.predict(x_teste)
previsoes_lgbm

In [313]:
previsoes_lgbm.shape

(276,)

In [314]:
for i in range(0,276):
  if previsoes_lgbm[i] >= .05:
     previsoes_lgbm[i] = 1
  else:
    previsoes_lgbm[i] = 0

In [ ]:
previsoes_lgbm

In [316]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_lgbm)*100.0))

print(confusion_matrix(y_teste, previsoes_lgbm))

print(classification_report(y_teste, previsoes_lgbm))

Acurácia: 64.49%
[[ 23  98]
 [  0 155]]
              precision    recall  f1-score   support

           0       1.00      0.19      0.32       121
           1       0.61      1.00      0.76       155

    accuracy                           0.64       276
   macro avg       0.81      0.60      0.54       276
weighted avg       0.78      0.64      0.57       276



In [ ]:
#Análise dados Treino (LightGBM)
previsoes_treino = lgbm.predict(x_treino)
previsoes_treino

In [318]:
previsoes_treino.shape

(641,)

In [319]:
for i in range(0,641):
  if previsoes_treino[i] >= .05:
     previsoes_treino[i] = 1
  else:
    previsoes_treino[i] = 0

In [ ]:
previsoes_treino

In [321]:
accuracy_score(y_treino, previsoes_treino)

0.6365054602184087